Batteries/Data/ByteSlice.lean

```diff
@@ -0,0 +1,91 @@
+
+/-
+Copyright (c) 2021 Mario Carneiro. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mario Carneiro
+-/
+
+set_option autoImplicit true
+
+namespace Nat
+
+/- Up -/
+
+/-- A well-ordered relation for "upwards" induction on the natural numbers up to some bound `ub`. -/
+def Up (ub a i : Nat) := i < a ∧ i < ub
+
+theorem Up.next {ub i} (h : i < ub) : Up ub (i+1) i := ⟨Nat.lt_succ_self _, h⟩
+
+theorem Up.WF (ub) : WellFounded (Up ub) :=
+  Subrelation.wf (h₂ := (measure (ub - ·)).wf) fun ⟨ia, iu⟩ => Nat.sub_lt_sub_left iu ia
+
+/-- A well-ordered relation for "upwards" induction on the natural numbers up to some bound `ub`. -/
+def upRel (ub : Nat) : WellFoundedRelation Nat := ⟨Up ub, Up.WF ub⟩
+
+end Nat
+
+set_option linter.missingDocs false in -- TODO
+/-- A terminal byte slice, a suffix of a byte array. -/
+structure ByteSliceT := (arr : ByteArray) (off : Nat)
+
+namespace ByteSliceT
+
+/-- The number of elements in the byte slice. -/
+@[inline] def size (self : ByteSliceT) : Nat := self.arr.size - self.off
+
+/-- Index into a byte slice. The `getOp` function allows the use of the `buf[i]` notation. -/
+@[inline] def getOp (self : ByteSliceT) (idx : Nat) : UInt8 := self.arr.get! (self.off + idx)
+
+end ByteSliceT
+
+/-- Convert a byte array into a terminal slice. -/
+def ByteArray.toSliceT (arr : ByteArray) : ByteSliceT := ⟨arr, 0⟩
+
+set_option linter.missingDocs false in -- TODO
+/-- A byte slice, given by a backing byte array, and an offset and length. -/
+structure ByteSlice := (arr : ByteArray) (off len : Nat)
+
+namespace ByteSlice
+
+/-- Convert a byte slice into an array, by copying the data if necessary. -/
+def toArray : ByteSlice → ByteArray
+  | ⟨arr, off, len⟩ => arr.extract off len
+
+/-- Index into a byte slice. The `getOp` function allows the use of the `buf[i]` notation. -/
+@[inline] def getOp (self : ByteSlice) (idx : Nat) : UInt8 := self.arr.get! (self.off + idx)
+
+
+/-- The inner loop of the `forIn` implementation for byte slices. -/
+def forIn.loop [Monad m] (f : UInt8 → β → m (ForInStep β))
+    (arr : ByteArray) (off _end : Nat) (i : Nat) (b : β) : m β :=
```

Comment:
TODO: Efficient implementation as in core.

Interestingly, core's `ByteArray.foldl` has optional start/stop parameters but `ByteArray.forIn` does not. It would be great if it did.

Reply:
`off` is redundant but `_end` isn't, I think? It's used in the `if` statement. (Note that this inner function takes a `ByteArray`, not a `ByteSlice`. This is an optimization.)